accepted/00000-saltclass-external-pillar.md

```diff
@@ -0,0 +1,147 @@
+
+## Step 2: Switching necessary parts to use saltclass
+
+As a necessary parts are considered image pillars and Formulas with Forms integration.
+Image pillars because of https://github.com/SUSE/spacewalk/issues/8337 and Formulas with Forms integration because of need to remove code duplication, solve some corner cases when data merging did not provide correct values ( https://github.com/SUSE/spacewalk/issues/8367 ), yomi enablement if implemented ( https://github.com/SUSE/spacewalk/issues/8367 ) and can lead to simplification of saltboot mechanism as well.
+
+The major motivator in these cases is the jinja preparsing of the pillar data.
+
+### Changes in Formulas with Forms implementation
+
+1) Make FormulaFactory.java to store not just raw `json` data, but complete pillar structure as `saltclass pillar data`
+   This formula pillar will be stored as a `saltclass` `class` data and then included from `saltclass pillar data `of minion (this step may conflict with below mentioned Step 2b about migrating `existing pillar data`)
+2) If `saltclass` is also enabled as a tops module, minion `saltclass pillar data` can contain a list of state names to be included in the highstate. If `saltclass` is not enabled as a tops module then I recommend to keep this functionality in the `suma_minion.py` for this round of implementation.
+
+> **NOTE**
+>Choosing `saltclass` to be also enabled as a tops module and using it to load formulas can make implementation of https://github.com/SUSE/spacewalk/issues/9490 and https://github.com/SUSE/spacewalk/issues/2714 considerably harder.
+
+3) Adapt FormulaFactory.java to load data for formula directly from pillars instead of using custom `json` file.
+
+### Changes for image pillars are
+
+1) Adapt SaltStateGeneratorService.java to store pillar data as a `saltclass` `class` and enable it for all minions
+2) Migrate existing image pillars to saltclass classes
+
+## Step 3: Migrating existing pillar data to use satlclass
+
+Migrating existing pillar data will require two steps:
+1) Modification of java code storing minion pillars:
+  - drop `pillar_` prefix from pillar files
+  - store pillar data under `pillar` key
+2) Rename existing pillar files from `pillar_{minion_id}.yml` to `{minion_id}.yml`
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+## Big refactoring underway, but stalled
+
+There is an upstream pull request about major `saltclass` refactoring - https://github.com/saltstack/salt/pull/52407 - however this PR seems to be stalled. This PR meant to fix several issues and fulfil couple of feature requests.
+The unknown is if the PR will be ever finished or if we can invest and try to finalize it ourselves. It is possible that the work stopped only temporarily due to changes in salt releases, or that the author being external contributor chose to abandon the effort.
+
+## Performance impact
+
+Including another pillar engine can have a negative performance impact. (#TODO measure)
```

Comment:
That's why I'd go all in with saltclass or not at all. Of course we can take a detour during development in Uyuni, but once we hit a major release milestone we should have just one of the two pillar plugins.
Performance-wise, one thing that could potentially hit us hard is that JSON parsing is supposedly _much_ faster than YAML parsing, especially if there's Jinja involved as well. We need to measure that with very large data sets and also look at peak memory usage etc. when thousands of minions ask for their pillars.
If necessary, this can be partially fixed by some clever caching though.